code/modules/jobs/job_exp.dm

```diff
@@ -0,0 +1,274 @@
+	for(var/client/L in GLOB.clients)
+		if(L.inactivity >= 6000)
+			continue
+		addtimer(L.update_exp_list(mins, ann),10)
+		CHECK_TICK
+
+//Manual incrementing/updating
+/client/proc/update_exp_client(var/minutes, var/announce_changes = 0)
+	if(!src ||!ckey || !config.use_exp_tracking)
+		return
+	if(!SSdbcore.Connect())
+		return -1
+	var/datum/DBQuery/exp_read = SSdbcore.NewQuery("SELECT job, minutes FROM [format_table_name("role_time")] WHERE ckey = '[sanitizeSQL(ckey)]'")
+	if(!exp_read.Execute())
+		var/err = exp_read.ErrorMsg()
+		log_game("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		message_admins("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		return
+	var/list/play_records = list()
+	while(exp_read.NextRow())
+		play_records[exp_read.item[1]] = text2num(exp_read.item[2])
+
+	for(var/rtype in GLOB.exp_jobsmap)
+		if(!play_records[rtype])
+			play_records[rtype] = 0
+	if(mob.stat == CONSCIOUS && mob.mind.assigned_role)
+		play_records[EXP_TYPE_LIVING] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Living EXP!")
+		for(var/category in GLOB.exp_jobsmap)
+			if(GLOB.exp_jobsmap[category]["titles"])
+				if(mob.mind.assigned_role in GLOB.exp_jobsmap[category]["titles"])
+					play_records[category] += minutes
+					if(announce_changes)
+						to_chat(mob,"<span class='notice'>You got: [minutes] [category] EXP!")
+		if(mob.mind.special_role)
+			play_records[EXP_TYPE_SPECIAL] += minutes
+			if(announce_changes)
+				to_chat(mob,"<span class='notice'>You got: [minutes] Special EXP!")
+	else if(isobserver(mob))
+		play_records[EXP_TYPE_GHOST] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Ghost EXP!")
+	else if(minutes)	//Let "refresh" checks go through
+		return
+	prefs.exp = play_records
+
+	for(var/jtype in play_records)
+		var jobname = jtype
+		var time = play_records[jtype]
+		var/datum/DBQuery/update_query = SSdbcore.NewQuery("INSERT INTO [format_table_name("role_time")] (`ckey`, `job`, `minutes`) VALUES ('[sanitizeSQL(ckey)]', '[sanitizeSQL(jobname)]', '[sanitizeSQL(time)]') ON DUPLICATE KEY UPDATE minutes = '[sanitizeSQL(time)]'")
+		if(!update_query.Execute())
+			var/err = update_query.ErrorMsg()
+			log_game("SQL ERROR during exp_update_client update. Error : \[[err]\]\n")
+			message_admins("SQL ERROR during exp_update_client update. Error : \[[err]\]\n")
+			return
+
+
+
+/client/proc/update_exp_list(var/minutes, var/announce_changes = 0)
+	if(!src ||!ckey || !config.use_exp_tracking)
+		return
+
+	var/list/play_records = prefs.exp
+	var/list/old_records = prefs.exp.Copy()
+
+	for(var/rtype in GLOB.exp_jobsmap)
+		if(!play_records[rtype])
+			play_records[rtype] = 0
+	if(mob.stat == CONSCIOUS && mob.mind.assigned_role)
+		play_records[EXP_TYPE_LIVING] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Living EXP!")
+		for(var/category in GLOB.exp_jobsmap)
```

Comment:
See, this isn't how you so should do this.

Don't add the exp to the different categories, save the exp by job/type, then have checking the exp add up the relevant jobs/types.

As it is, if you make a new exp requirement, it wouldn't track. if you just track all jobs, as one, you don't have to worry about that down the line. (eg, say you add a new job, call it the troll, and make it require clown and/or mime exp, well it would have no data under this system, but if you just tracked all jobs and roles, regardless, new roles and requirements down the line would have data) (Also @nfreader would love the extra data even if no jobs require that job)